### PR TITLE
feat: Allow adding newlines in description for tasks

### PR DESCRIPTION
**Pull request by akinwale**
### Details
Make the text input controls for the task descriptions support multiline input and autogrow.

### Fixed Issues
$ https://github.com/Expensify/App/issues/19135
PROPOSAL: https://github.com/Expensify/App/issues/19135#issuecomment-1551880839

### Tests
1. Open the Expensify app or web
2. Select Assign task from the global "New" menu
3. Attempt to enter multiple lines when providing a description
4. Click Next
5. Click on Description
6. Verify that you can enter or edit multiple lines
7. After saving the task, navigate to the created task.
8. Click on the description and verify that you can edit or enter multiple lines.

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
Same as tests.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/4160319/0bcb3fe2-d2ab-4be2-bcca-2da3fd1dee66

</details>

<details>
<summary>Mobile Web - Chrome</summary>

[19135-android-chrome.webm](https://github.com/Expensify/App/assets/4160319/18defcfa-18d1-4256-8c67-622f3bcda8ed)

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/4160319/0f0826a0-58fa-4fcd-8fcf-98fd0d1699a2

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/4160319/e8221e46-c4d8-4d92-84fa-2610e0a4f97d

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/4160319/744e1ff2-105d-4e06-9393-21a206752816

</details>

<details>
<summary>Android</summary>

[19135-android-native.webm](https://github.com/Expensify/App/assets/4160319/da604f28-8931-4a95-a017-13d27079e0f2)

</details>

### CODE REVIEW

**Comment by akinwale** [comment]
@sobitneupane There appears to be a couple of bugs with the way the autogrow TextInput control is implemented. The `welcomeMessage` input in the `WorkspaceInviteMessagePage` component also exhibits these issues, but it seems to go unnoticed because it is prefilled with four lines of text, and pressing Enter does not do anything when you try to edit the invite message.

1. When the control initially loads and it is blank, the input area expands to fill the max height. However, once you start typing, the input area height auto-adjusts to fit the single line of text that the user is currently on. This happens on the following platforms: Safari (iOS and macOS), desktop, Android Chrome.
2. When you press Enter to create a new line, the input area does not auto adjust to the next line until the user starts typing characters on the new line. This seems to automatically resolve itself once the control reaches the container max height. This happens on the following platforms: Web (Safari, Chrome), iOS Safari, Android Chrome, Android native.

Are these known issues? If not, shall I report them on Slack?

**Comment by sobitneupane** [comment]
@akinwale I don't think we would expect you to solve those issues if they are pre-existing and will only be re-surfaced by your PR.

cc: @youssef-lr

**Review by sobitneupane — COMMENTED** [review]
Cursor is expected at the end of already input text.

https://github.com/Expensify/App/assets/25876548/2598d0d2-4793-4448-ae82-198d07f1f242

**Comment by akinwale** [comment]
> @akinwale Instead of trimming the description, we should continue description from the next line.

@sobitneupane This would require modifying the `MenuItem` component by changing how the title is displayed. Right now, it is set to `numberOfLines={1}`. I can add a boolean prop called `multilineTitle` which can be used to remove the number of lines, and wrap the text as required. These seem to be out of scope for this PR considering the number of changes to a different component which is widely used across the app, but I can go ahead if necessary. Please let me know what you think.

https://github.com/Expensify/App/blob/b424edcbe7cea7ed5fe36bc5602f6eb4ed1d78bf/src/components/MenuItem.js#L159-L166

**Comment by akinwale** [comment]
> @akinwale Can you please confirm [this](https://github.com/Expensify/App/pull/19377#issuecomment-1559280964)?

@sobitneupane Yeah, this is done. Please check the latest commit.

**Comment by sobitneupane** [comment]
Since we have made the text input multiline, I think we should make the Description text multiline.
Screenshot 2023-05-24 at 11 50 59

Something like this:
Screenshot 2023-05-24 at 11 53 01

But the lower section of the report is scrollable. So, we should either make the whole report scrollable including Task Title and Task Description or fix the height of Description and make just the Description scrollable.

https://github.com/Expensify/App/assets/25876548/1cbd23c3-6d7e-4906-8a85-2cfcbe5b9b4e



@youssef-lr What is your opinion on this?

**Comment by sobitneupane** [comment]
@akinwale  https://github.com/Expensify/App/pull/19377#issuecomment-1560512162
Sorry, I don't understand what you mean? Do you mean that with your latest commit the issue no longer exists? Or Are you confirming that it cannot be solved frontend?

**Comment by akinwale** [comment]
> User cannot save description with newline. It might be a backend issue. @akinwale Can you please look around the code and confirm it is not a front end issue.

Sorry, I misunderstood. I thought you were referring to the multiline description here.

Not being able to save description with newline is a backend issue. I've confirmed this as descriptions without newlines are saved just fine. I was planning to submit a bug report for this after this issue got merged.

**Comment by sobitneupane** [comment]
@youssef-lr User cannot save description with newline. Probably it is a backend issue.

https://github.com/Expensify/App/assets/25876548/f9cd7f41-0c8e-41df-a1da-769314dd79c1

**Comment by akinwale** [comment]
> But the lower section of the report is scrollable. So, we should either make the whole report scrollable including Task Title and Task Description or fix the height of Description and make just the Description scrollable.

@sobitneupane I think the intended design here is for the header to be fixed, while the report below is scrollable. You can also observe the same behaviour with the `MoneyRequestHeader` component for money / IOU requests.

Will wait for @youssef-lr's thoughts on this.

**Comment by akinwale** [comment]
@sobitneupane I think you could also start a discussion on Slack about this so that we can have an idea on the design and close this soon.

**Comment by sobitneupane** [comment]
@youssef-lr Bump on https://github.com/Expensify/App/pull/19377#issuecomment-1560530638 and https://github.com/Expensify/App/pull/19377#issuecomment-1560518276

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/pull/19377#issuecomment-1562549362

@akinwale Sounds good. Can you please start a discussion on #expensify-open-source?

**Comment by akinwale** [comment]
> [#19377 (comment)](https://github.com/Expensify/App/pull/19377#issuecomment-1562549362)
> 
> @akinwale Sounds good. Can you please start a discussion on #expensify-open-source?

@sobitneupane Based on @thienlnam's reply in the Slack discussion, looks like we can move forward with this as it is right now (no further design changes required). I have fixed the merged conflicts against the latest main branch.

**Comment by sobitneupane** [comment]
Thanks @akinwale for carrying [the discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1685087242060519) in slack. 

@youssef-lr Should we hold this PR for https://github.com/Expensify/App/pull/19377#issuecomment-1560530638?

**Comment by thienlnam** [comment]
I can create another issue to track the BE issue. But we can still move forward on this - you can still have a long description without new lines

**Comment by sobitneupane** [comment]
Thanks @thienlnam.

@akinwale Can you please resolve the conflict and let me know. If everything goes well, let's get it merged today.

**Comment by akinwale** [comment]
> Thanks @thienlnam.
> 
> @akinwale Can you please resolve the conflict and let me know. If everything goes well, let's get it merged today.

@sobitneupane Will do.

Also, @tjferriss raised a point about merging this issue with one for the money request header [here](https://github.com/Expensify/App/issues/19872#issuecomment-1574403080) because they are very similar. Do you want to clarify if it's ok to move forward with this?

**Comment by sobitneupane** [comment]
[Commented](https://github.com/Expensify/App/issues/19872#issuecomment-1576358312) on the issue

**Comment by akinwale** [comment]
> Commented

@sobitneupane Clarified [here](https://github.com/Expensify/App/issues/19872#issuecomment-1577314369) so we can move forward with this PR as is. I have fixed the merge conflicts so I'll just wait for the final review. Thanks.

**Comment by sobitneupane** [comment]
@akinwale Android app crashes on opening task report with multiline description. 

Please make sure to retest PR on all platform after making any major update and please do update the screenshots in PR body as well.

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>
Screenshot 2023-06-06 at 14 27 36
Screenshot 2023-06-06 at 14 27 27
Screenshot 2023-06-06 at 14 29 22
Screenshot 2023-06-06 at 14 32 59

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/25876548/ae68253c-e2f7-4f56-8b39-d5c89c5cd06e

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/25876548/3e452b9e-3dca-48c6-82bb-f21198be5932

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/25876548/99c2bce3-da7b-491a-ae14-34b5ef35c64e


</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/25876548/1b24dec6-f6cb-4fcc-8220-6bdbc363a747

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/25876548/fb3ffa74-5443-4038-9f96-f73147d2afb0

</details>

**Comment by sobitneupane** [comment]
In addition to [this](https://github.com/Expensify/App/pull/19377#issuecomment-1578262321), as suggested by @thienlnam in slack, let's not change anything in task report and continue with the current design. I think leaving a single line description is okay for now. If we have multiline input, it will have [this issue](https://github.com/Expensify/App/pull/19377#issuecomment-1560518276).

> It’s fine to leave this current design in order to close this out - I’ll create a new issue for the redesign that can handle those issues. Thanks for investigating

File: src/components/MenuItem.js
Comment (sobitneupane) [inline src/components/MenuItem.js:82]: Looks like you have forgotten to undo the change here.

**Review by sobitneupane — COMMENTED** [review]
As suggested by @thienlnam in slack, we will leave the preview of task description as it is (i.e., single line) and will make the change only on task description input.

File: src/components/MenuItem.js
Comment (akinwale) [reply src/components/MenuItem.js]: @sobitneupane Oops. Fixed. Thanks.

**Review by sobitneupane — COMMENTED** [review]
Changes look good and tests well.

https://github.com/Expensify/App/assets/25876548/5c4bb6fc-e88b-433b-9625-96208fe5b366



@thienlnam @youssef-lr

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Review by youssef-lr — APPROVED** [review]
LGTM!

File: src/pages/tasks/TaskDescriptionPage.js
Comment (marcaaron) [inline src/pages/tasks/TaskDescriptionPage.js:60]: Should have been a custom hook maybe. We try to follow the DRY principle.

File: src/pages/tasks/TaskDescriptionPage.js
Comment (marcaaron) [inline src/pages/tasks/TaskDescriptionPage.js:51]: Who knows what this page might use in the future or if it will continue to exist at all. Referencing it here does not add anything.

File: src/pages/tasks/TaskDescriptionPage.js
Comment (marcaaron) [inline src/pages/tasks/TaskDescriptionPage.js:54]: Why disabled?

File: src/pages/tasks/TaskDescriptionPage.js
Comment (akinwale) [reply src/pages/tasks/TaskDescriptionPage.js:60]: Just saw these comments, sorry. Yeah, this would probably work but I'll be removing it anyway due to the React Native iOS bug with the selection prop.

File: src/pages/tasks/TaskDescriptionPage.js
Comment (akinwale) [reply src/pages/tasks/TaskDescriptionPage.js:54]: I didn't think doing an early return here made much sense. It would've been 3 additional lines of code `if (!condition) { return; }` which don't really add anything.

**Pull request merged**
